src/handler.js

```diff
@@ -0,0 +1,330 @@
+function createLogReference(queue, taskId, run) {
+  let logUrl = queue.buildUrl(
+    queue.getArtifact, taskId, run.runId, 'public/logs/live_backing.log'
+  );
+
+  return {
+    // XXX: This is a magical name see 1147958 which enables the log viewer.
+    name: 'builds-4h',
+    url: logUrl
+  };
+}
+
+// Filters the task routes for the treeherder specific route.  Once found,
+// the route is parsed into distinct parts used for constructing the
+// Treeherder job message.
+function parseRouteInfo(prefix, taskId, routes) {
+  let route = routes.filter((r) => {
+    return r.split('.')[0] === prefix;
+  });
+
+  if (route.length != 1) {
+    throw new Error(
+      `Could not determine treeherder route.  Either there is no route, ` +
+      `or more than one matching route exists.  Task ID: ${taskId} Routes: ${routes}`
+    );
+  }
+
+  return parseRoute(route[0]);
+}
+
+
+function validateTask(validate, taskId, task, schema) {
+    if (!task.extra || !task.extra.treeherder) {
+      throw new Error(`Message is missing Treeherder job configuration. Task ID: ${taskId}`);
+    }
+    let validationErrors = validate(task.extra.treeherder, schema);
+    if (validationErrors) {
+      throw new Error(`Message contains an invalid Treeherder job configuration. Task ID: ${taskId} ${validationErrors}`);
+    }
+}
+
+export class Handler extends events.EventEmitter {
+  constructor(options) {
+    super();
+    this.queueEvents = options.queueEvents;
+    this.queue = options.queue;
+    this.scheduler = options.scheduler;
+    this.listener = options.listener;
+    this.prefix = options.prefix;
+    this.publisher = options.publisher;
+    this.validator = options.validator;
+  }
+
+  // Starts up the message handler and listens for messages
+  async start() {
+    this.listener.on('message', async (message) => {
+      await this.handleMessage(message).catch(err => {
+        console.log(`Error caught when processing message. ${err.message}. ${err.stack}`);
+      });
+    });
+    await this.listener.resume();
+  }
+
+  // Listens for Task event messages and invokes the appropriate handler
+  // for the type of message received.
+  //
+  // Only messages that contain the properly formatted routing key and contains
+  // treeherder job information in task.extra.treeherder are accepted
+  async handleMessage(message) {
+    let taskId = message.payload.status.taskId;
+    let parsedRoute = parseRouteInfo(this.prefix, taskId, message.routes);
+
+    let task = await this.queue.task(taskId);
+    validateTask(this.validator, taskId, task, TASK_TREEHERDER_SCHEMA);
+
+    switch (message.exchange) {
+      case this.queueEvents.taskPending().exchange:
+        let runId = message.payload.runId;
+        let run = message.payload.status.runs[message.payload.runId];
+        // If the task run was created for an infrastructure rerun, then resolve
+        // the previous run as retried.
+        if (taskIsRerun(runId, run)) {
```

Comment:
Let's rename this...to not say rerun or retry... it has to do with reporting the previous run.